File: 375. Guess Number Higher or Lower II/GetMoneyAmount.py
class Solution:
    def findMinMoney(self, lo, hi, memo) -> int:
        
        if lo >= hi:
            return 0
        
        if (lo, hi) in memo:
            return memo[(lo, hi)]
        
        min_money = float('inf')
        mid = (lo+hi)//2
        
        for num in range(lo, hi+1):
            
            money = num + max(self.findMinMoney(lo, num-1, memo), self.findMinMoney(num+1, hi, memo))
            min_money = min(min_money, money)
        
        memo[(lo, hi)] = min_money
        return int(min_money)
    
    
    def getMoneyAmount(self, n: int) -> int:
        
        return self.findMinMoney(1, n, {})


sol = Solution()
print(sol.getMoneyAmount(10))

File: 375. Guess Number Higher or Lower II/GetMoneyAmount.rs
use std::cmp;
use std::collections::HashMap;
// use std::collections::tuple;

impl Solution {
    
    pub fn find_min_money(lo: i32, hi: i32, memo: &mut HashMap<(i32, i32), i32>) -> i32 {
        
        if (lo >= hi) {
            return 0;
        }
        
        if memo.contains_key(&(lo, hi)) {
            return *memo.get(&(lo, hi)).unwrap_or(&0);
        }
        
        let mut min_money = i32::MAX;
        
        for num in lo..hi+1 {
            
            let money = num + cmp::max(Solution::find_min_money(lo, num-1, memo), Solution::find_min_money(num+1, hi, memo));
            min_money = cmp::min(min_money, money);
            // println!("{}", money);
        }
        
        memo.insert((lo, hi), min_money);
        return min_money;
        
    }
    
    pub fn get_money_amount(n: i32) -> i32 {
        
        let mut memo = HashMap::new();
        
        return Solution::find_min_money(1, n, &mut memo);
        
    }
}